### PR TITLE
SONARCS-831: Do not overload 'operator ==' on reference types

### PR DESCRIPTION
**Pull request by Evangelink**

### CODE REVIEW

File: src/SonarAnalyzer.CSharp/Rules/DotNotOverloadOperatorEqual.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            var classDeclaration = declaration.Parent as ClassDeclarationSyntax;
+            if (classDeclaration == null)
+            {
+                return;
+            }
+
+            var hasAdditionOrSubstractionOverload =
+                classDeclaration.ChildNodes()
+                                .OfType<OperatorDeclarationSyntax>()
+                                .Any(op => op.OperatorToken.IsKind(SyntaxKind.PlusToken) ||
```
Comment (valhristov) [inline src/SonarAnalyzer.CSharp/Rules/DotNotOverloadOperatorEqual.cs]: Don't we have something like op.IsKind(SyntaxKind.PlusOperator), e.g. on the operator, not the token? The operator check seems more robust than the token.

File: src/Tests/SonarAnalyzer.UnitTest/TestCases/DotNotOverloadOperatorEqual.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
+using System;
+using System.Collections.Generic;
+
+namespace Tests.Diagnostics
+{
+    class MyType
+    {
+        public static MyType operator ==(MyType x, MyType y) // Noncompliant
```
Comment (valhristov) [inline src/Tests/SonarAnalyzer.UnitTest/TestCases/DotNotOverloadOperatorEqual.cs]: I cannot find tests for the diagnostic message.

File: src/SonarAnalyzer.CSharp/Rules/DotNotOverloadOperatorEqual.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        private void CheckForIssue(SyntaxNodeAnalysisContext analysisContext)
+        {
+            var declaration = (OperatorDeclarationSyntax)analysisContext.Node;
+
+            if (!declaration.OperatorToken.IsKind(SyntaxKind.EqualsEqualsToken))
+            {
+                return;
+            }
+
+            var structDeclaration = declaration.Parent as StructDeclarationSyntax;
```
Comment (valhristov) [inline src/SonarAnalyzer.CSharp/Rules/DotNotOverloadOperatorEqual.cs]: this is minor, but you could replace this with:
if (declaration.Parent is StructDeclarationSyntax)
{
    return;
}

File: src/SonarAnalyzer.CSharp/RspecStrings.resx [hunk 4]
```diff
@@ -1,17 +1,17 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <root>
-  <!--
```
Comment (valhristov) [inline src/SonarAnalyzer.CSharp/RspecStrings.resx]: Why is the header changed?

Comment (Evangelink) [inline src/SonarAnalyzer.CSharp/RspecStrings.resx]: No idea I just updated the rule as the description changed.

File: src/SonarAnalyzer.CSharp/Rules/DotNotOverloadOperatorEqual.cs (hunk 1 above)
Comment (Evangelink) [inline src/SonarAnalyzer.CSharp/Rules/DotNotOverloadOperatorEqual.cs]: I haven't found anything better.

File: src/Tests/SonarAnalyzer.UnitTest/TestCases/DotNotOverloadOperatorEqual.cs (hunk 2 above)
Comment (Evangelink) [inline src/Tests/SonarAnalyzer.UnitTest/TestCases/DotNotOverloadOperatorEqual.cs]: Done

**Pull request merged**